versions/versions.go

```diff
@@ -101,48 +97,91 @@ type Extraction struct {
+ * must match each corresponding sub-directories and file name for the path to
+ * be retained.
+ *
+ * The function walk the file tree stored in the S3 bucket `bucketName` and
+ * collect the full paths that match `regex` along the way. It takes care of
+ * following only the branches (prefix in S3 terms) that match with the
+ * corresponding section of `regex`.
+ */
+func GetMatchingPathsFromBucket(client s3resource.S3Client, bucketName string, regex string) ([]string, error) {
+	type work struct {
+		prefix  string
+		remains []string
+	}
 
-	for _, section := range sections {
-		if re.MatchString(section) {
-			validSections = append(validSections, nonRE.ReplaceAllString(section, "${chr}"))
+	specialCharsRE := regexp.MustCompile(`[\\\*\.\[\]\(\)\{\}\?\|\^\$\+]`)
+
+	matchingPaths := []string{}
+	queue := []work{{prefix: "", remains: strings.Split(regex, "/")}}
+	for len(queue) != 0 {
+		prefix := queue[0].prefix
+		remains := queue[0].remains
+		section := remains[0]
+		remains = remains[1:]
+		queue = queue[1:]
+		if !specialCharsRE.MatchString(section) && len(remains) != 0 {
+			// No special char so it can match a single string and we can just extend the prefix
+			// but only if some remains exists, i.e. the section is not a leaf.
+			prefix += section + "/"
+			queue = append(queue, work{prefix: prefix, remains: remains})
+			continue
+		}
+		// Let's list what's under the current prefix and see if that matches with the section
+		var prefixRE *regexp.Regexp
+		if len(remains) != 0 {
+			// We need to look deeper so full prefix will end with a /
+			prefixRE = regexp.MustCompile(prefix + section + "/")
 		} else {
-			break
+			prefixRE = regexp.MustCompile(prefix + section)
+		}
+		var (
+			continuationToken *string
+			truncated         bool
+		)
+		for continuationToken, truncated = nil, true; truncated; {
+			s3ListChunk, err := client.ChunkedBucketList(bucketName, prefix, continuationToken)
+			if err != nil {
+				return []string{}, err
+			}
+			truncated = s3ListChunk.Truncated
+			continuationToken = s3ListChunk.ContinuationToken
+
+			if len(remains) != 0 {
+				// We need to look deeper so full prefix will end with a /
+				for _, commonPrefix := range s3ListChunk.CommonPrefixes {
+					if prefixRE.MatchString(commonPrefix) {
+						queue = append(queue, work{prefix: commonPrefix, remains: remains})
+					}
+				}
+			} else {
+				// We're looking for a leaf
+				for _, path := range s3ListChunk.Paths {
+					if prefixRE.MatchString(path) {
+						matchingPaths = append(matchingPaths, path)
+					}
+				}
+			}
 		}
 	}
-
-	if len(validSections) == 0 {
-		return ""
-	}
-
-	return strings.Join(validSections, "/") + "/"
+	return matchingPaths, nil
 }
 
 func GetBucketFileVersions(client s3resource.S3Client, source s3resource.Source) Extractions {
-	regexp := source.Regexp
-	hint := PrefixHint(regexp)
+	regex := source.Regexp
 
-	paths, err := client.BucketFiles(source.Bucket, hint)
```

Comment:
Is `BucketFiles` in used anymore?

Reply:
It is still in used by `integration/out_test.go` to retrieve the content of the test bucket and validate the out operation worked as expected.

`BucketFiles` is built on top of `getBucketContents` mentioned above. It can certainly be rewritten to use `ChunkedBucketList` instead which would allow the removal of `getBucketContents` completely and so any reference to the old `ListObjects`.

I will give it a try so stay tuned. :-)

Reply:
First, sorry for the force-pushed rebase, I forgot again to pass `--signoff` to git :disappointed:

I've pushed two more commits.
- First one replaces the `BucketFiles` implementation based on `getBucketContents` with one using `ChunkedBucketList` and removes `getBucketContents` entirely.
- Second one adds the ability to skip the test "with a large file that is multiple of MaxUploadParts". With my bandwidth it estimates the upload would take >45 minutes.

Using `ChunkedBucketList` for `BucketFiles` exposed a bug in its implementation. As I too happily squashed it with the initial implementation I'd like to reproduce the diff here:
```patch
@@ -201,11 +226,11 @@ func (client *s3client) ChunkedBucketList(bucketName string, prefix string, cont
        }
        commonPrefixes := make([]string, len(response.CommonPrefixes))
        paths := make([]string, len(response.Contents))
-       for _, commonPrefix := range response.CommonPrefixes {
-               commonPrefixes = append(commonPrefixes, *commonPrefix.Prefix)
+       for idx, commonPrefix := range response.CommonPrefixes {
+               commonPrefixes[idx] = *commonPrefix.Prefix
        }
-       for _, path := range response.Contents {
-               paths = append(paths, *path.Key)
+       for idx, path := range response.Contents {
+               paths[idx] = *path.Key
        }
        return BucketListChunk{
                Truncated:         *response.IsTruncated,
```
The issue was that both arrays `commonPrefixes` and `paths` were pre-allocated to the known size of the values received from the S3 API but then the values where appended to it instead of being copied into the arrays. As the consequence, the arrays were twice as large as needed with half of them being empty strings. This was not noticed before because the check operation then filters the resulting values against the regexp and the empty strings were all discarded at that time. As `BucketFiles` does not do any filtering the issue jumped out.

As I'm pretty much a newbie in Go, I wonder if there is any better implementation.

Reply:
@odormond this is a quick comment before I dig into the code, I might have misunderstood:

The problem might simply be that you have two ways to preallocate a slice: preallocating the capacity or the length.
If preallocating the length, then the proper way to iterate is to use an index. If on the other hand the slice is preallocated on the capacity, then the proper way to iterate is with append, which will operate without further allocations.

I suggest to read https://go.dev/blog/slices-intro, which will explain in detail and will also show that Note that `commonPrefixes` and `paths` are "slices" and not "arrays".

Reply:
Thanks @marco-m-pix4d! Nice blog post with a clear explanation.

So alternatively this can be fixed by setting the length to 0 and the capacity to the target value (tested):
```patch
@@ -199,8 +224,8 @@ func (client *s3client) ChunkedBucketList(bucketName string, prefix string, cont
        if err != nil {
                return BucketListChunk{}, err
        }
-       commonPrefixes := make([]string, len(response.CommonPrefixes))
-       paths := make([]string, len(response.Contents))
+       commonPrefixes := make([]string, 0, len(response.CommonPrefixes))
+       paths := make([]string, 0, len(response.Contents))
        for _, commonPrefix := range response.CommonPrefixes {
                commonPrefixes = append(commonPrefixes, *commonPrefix.Prefix)
        }
```

In any case the loops are still needed due to the pointer dereferences which prevent the use of `copy`. So I don't think there is anything more I can do here.